imagersite/imager_images/tests.py

```diff
@@ -87,3 +79,115 @@ def test_add_photos_to_albums(self):
+        album = Album.objects.all()[0]
+        c = Client()
+        response = c.get('/images/album/{}/'.format(album.id), follow=True)
+        self.assertEqual(len(response.redirect_chain), 1)
+        self.assertIn('form method="post" action="/login/"',
+                      response.content)
+
+
+class PhotoViewTestCase(TestCase):
+    def setUp(self):
+        user = UserFactory.create(username='userbob')
+        user.set_password('secret')
+        user.save()
+        photo = PhotoFactory.create(user=user)
+        photo.save()
+
+    def test_photo_detail_view(self):
+        photo = Photo.objects.all()[0]
+        c = Client()
+        c.login(username='userbob', password='secret')
+        response = c.get('/images/photos/{}/'.format(photo.id))
+        self.assertIn(photo.title, response.content)
+        self.assertIn(photo.description, response.content)
+
+    def test_photo_not_owner(self):
+        user = UserFactory.create(username='usereve')
+        user.set_password('secret')
+        user.save()
+        photo = Photo.objects.all()[0]
+        c = Client()
+        c.login(username='usereve', password='secret')
+        response = c.get('/images/photos/{}/'.format(photo.id))
+        assert response.status_code == 403
+
+    def test_photo_unauthenticated(self):
+        photo = Photo.objects.all()[0]
+        c = Client()
+        response = c.get('/images/photos/{}/'.format(photo.id), follow=True)
+        self.assertEqual(len(response.redirect_chain), 1)
+        self.assertIn('form method="post" action="/login/"',
+                      response.content)
+
+
+class LibraryViewTestCase(TestCase):
+    def setUp(self):
+        user = UserFactory.create(username='userbob')
+        user.set_password('secret')
+        user.save()
+        cover = PhotoFactory.create(user=user)
+        cover.save()
+        album = AlbumFactory.create(cover=cover, user=user)
+        album.save()
+        album.photos.add(cover)
+
+    def test_library_view(self):
+        photo = Photo.objects.all()[0]
+        album = Album.objects.all()[0]
+        c = Client()
+        c.login(username='userbob', password='secret')
+        response = c.get('/images/library/')
+        self.assertIn(photo.title, response.content)
+        self.assertIn(album.title, response.content)
+
+    def test_library_view_different_user(self):
+        photo = Photo.objects.all()[0]
+        album = Album.objects.all()[0]
+        c = Client()
+        c.login(username='usereve', password='secret')
+        response = c.get('/images/library/')
+        self.assertNotIn(photo.title, response.content)
+        self.assertNotIn(album.title, response.content)
+
+    def test_library_unauthenticated(self):
+        c = Client()
+        response = c.get('/images/library/', follow=True)
+        self.assertEqual(len(response.redirect_chain), 1)
+        self.assertIn('form method="post" action="/login/"',
+                      response.content)
```

Comment:
Perhaps more authentication related tests are needed, such as a user trying to add a photo to an album they don't own:

```python
    def test_other_users_cant_add_to_album(self):
        with self.assertRaises(AttributeError):
            self.bobalbum.add_photo(self.alicephoto)
```
It may be impossible for an attacker to reach this functionality from a url, but they may be able to find a back door.